src/data/rat/basic.lean

```diff
@@ -627,3 +627,79 @@ begin
 end
 
 end rat
+
+section meta_fns
+
+/--
+`rat.mk_numeral q` embeds `q` as a numeral expression inside a type with 0, 1, +, -, and /
+
+`type`: an expression representing the target type. This must live in Type 0.
+`has_zero`, `has_one`, `has_add`: expressions of the type `has_zero %%type`, etc.
+
+This function is similar to `expr.of_rat` but takes more hypotheses and is not tactic valued.
+ -/
+meta def rat.mk_numeral (type has_zero has_one has_add has_neg has_div : expr) : ℚ → expr
+| ⟨num, denom, _, _⟩ :=
+  let nume := num.mk_numeral type has_zero has_one has_add has_neg in
+  if denom = 1 then nume else
+    let dene := denom.mk_numeral type has_zero has_one has_add in
+    `(@has_div.div.{0} %%type %%has_div %%nume %%dene)
+
+/-- `rat.reflect q` represents the rational number `q` as a numeral expression of type `ℚ`. -/
+meta def rat.reflect : ℚ → expr :=
+rat.mk_numeral `(ℚ) `((by apply_instance : has_zero ℚ))
+         `((by apply_instance : has_one ℚ))`((by apply_instance : has_add ℚ))
+         `((by apply_instance : has_neg ℚ)) `(by apply_instance : has_div ℚ)
+
+section
+local attribute [semireducible] reflected
+meta instance : has_reflect ℚ := rat.reflect
+end
+
+/-- Evaluates an expression as a rational number,
+if that expression represents a numeral or the quotient of two numerals. -/
+protected meta def expr.to_pos_rat : expr → option ℚ
+| `(%%e₁ / %%e₂) := do m ← e₁.to_nat, n ← e₂.to_nat, some (rat.mk m n)
+| e              := do n ← e.to_nat, return (rat.of_int n)
+
+/-- Evaluates an expression as a rational number,
+if that expression represents a numeral, the quotient of two numerals,
+the negation of a numeral, or the negation of the quotient of two numerals. -/
+protected meta def expr.to_rat : expr → option ℚ
+| `(has_neg.neg %%e) := do q ← e.to_pos_rat, some (-q)
+| e                  := e.to_pos_rat
+
+/-- Evaluates an expression into a rational number, if that expression is built up from
+  numerals, +, -, *, /, ⁻¹  -/
+protected meta def expr.eval_rat : expr → option ℚ
```

Comment:
Is this function supposed to be recursive?

Reply:
Oops! I renamed it after Travis noticed the name clash with `to_rat`, which was originally defined in `norm_num.lean`.